### PR TITLE
Bug fixes in String builtin functions [TG-2459]

### PR DESCRIPTION
**Pull request by romainbrenguier**
These correct a few problems in builtin functions and create a new file for builtin functions.
These are commits extracted from https://github.com/diffblue/cbmc/pull/1947 which correct existing bugs.

### CODE REVIEW

File: src/solvers/refinement/string_refinement_util.cpp [hunk 1]
```diff
@@ -396,8 +396,19 @@ void string_dependenciest::for_each_successor(
...
+      {
+        const auto current = stack.back();
+        stack.pop_back();
+        if(const auto if_expr = expr_try_dynamic_cast<if_exprt>(current.get()))
+        {
+          stack.emplace_back(if_expr->true_case());
+          stack.emplace_back(if_expr->false_case());
+        }
+        else if(const auto node = node_at(to_array_string_expr(current)))
+            f(nodet(*node));
```
Comment (allredj) [inline src/solvers/refinement/string_refinement_util.cpp]: So what's the rule, 2-space or 4-space indent?

File: src/solvers/refinement/string_refinement_util.cpp [hunk 2]
```diff
@@ -396,8 +396,19 @@ void string_dependenciest::for_each_successor(
     const auto &builtin = builtin_function_nodes[node.index];
     for(const auto &s : builtin->string_arguments())
     {
-      if(const auto node = node_at(s))
-        f(nodet(*node));
+      std::vector<std::reference_wrapper<const exprt>> stack({s});
+      while(!stack.empty())
+      {
+        const auto current = stack.back();
+        stack.pop_back();
```
Comment (allredj) [inline src/solvers/refinement/string_refinement_util.cpp:403]: The fact that `pop_back()` does not return the popped value is outrageous...

File: src/solvers/refinement/string_refinement_util.cpp [hunk 3]
```diff
@@ -396,8 +396,19 @@ void string_dependenciest::for_each_successor(
...
+        if(const auto if_expr = expr_try_dynamic_cast<if_exprt>(current.get()))
+        {
+          stack.emplace_back(if_expr->true_case());
+          stack.emplace_back(if_expr->false_case());
+        }
+        else if(const auto node = node_at(to_array_string_expr(current)))
+            f(nodet(*node));
+      }
```
Comment (allredj) [inline src/solvers/refinement/string_refinement_util.cpp]: `else  UNREACHABLE`?

**Review by allredj — APPROVED** [review]
I just have a slight doubt that all cases are treated in `for_each_successor`.

File: src/solvers/refinement/string_builtin_function.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  const std::vector<mp_integer> &array,
+  const array_typet &array_type);
+
+string_transformation_builtin_functiont::
+  string_transformation_builtin_functiont(
+    const std::vector<exprt> &fun_args,
+    array_poolt &array_pool)
+{
+  PRECONDITION(fun_args.size() > 2);
+  const auto arg1 = expr_checked_cast<struct_exprt>(fun_args[2]);
```
Comment (peterschrammel) [inline src/solvers/refinement/string_builtin_function.cpp:25]: Could you document the magic indices and offsets please? A brief layout of the vectors would help.

File: src/solvers/refinement/string_refinement_util.cpp (hunk 1 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_refinement_util.cpp]: I just forgot to run clang-format...

File: src/solvers/refinement/string_builtin_function.cpp (hunk 4 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_builtin_function.cpp:25]: @peterschrammel I now document the constructor in the header, hopefully that should be clearer.

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**